tests/basis_test.py

```diff
@@ -1,330 +1,1167 @@
-
-    assert numpy.isclose(f0.derivative([1,-0.5]),[0,0]).all()
-    assert numpy.isclose(f1.derivative([1,-0.5]),[1,1]).all()
-    assert numpy.isclose(f2.derivative([1,-0.5]),[4,-2]).all()
-
-
-def test_cheb_2nd_derivative():
-    """Test if the correct derivative is generated."""
-    u0 = ChebyshevSecondKind(0)
-    u1 = ChebyshevSecondKind(1)
-    u2 = ChebyshevSecondKind(2)
-    assert u0.derivative(1) == pytest.approx(0)
-    assert u1.derivative(1) == pytest.approx(2)
-    assert u2.derivative(1) == pytest.approx(8)
-
-    assert u0.derivative(-1) == pytest.approx(0)
-    assert u1.derivative(-1) == pytest.approx(2)
-    assert u2.derivative(-1) == pytest.approx(-8)
-
-    assert u0.derivative(0.5) == pytest.approx(0)
-    assert u1.derivative(0.5) == pytest.approx(2)
-    assert u2.derivative(0.5) == pytest.approx(4)
-
-    assert numpy.isclose(u0.derivative([1,-1,0.5]),[0,0,0]).all()
-    assert numpy.isclose(u1.derivative([1,-1,0.5]),[2,2,2]).all()
-    assert numpy.isclose(u2.derivative([1,-1,0.5]),[8,-8,4]).all()
-
-
-
-def test_cheb_call_invalid_input():
-    """Test call raises error if input is outside domain [-1,1]"""
-    f = ChebyshevFirstKind(4)
+
+import smolyay
+
+
+@pytest.mark.parametrize(
+    "cheb_function",
+    [
+        smolyay.basis.ChebyshevFirstKind,
+        smolyay.basis.ChebyshevSecondKind,
+    ],
+)
+def test_cheb_initial(cheb_function):
+    """test degrees returns correctly"""
+    f2 = cheb_function(2)
+    assert f2.degree == 2
+    assert isinstance(f2.degree, int)
+    assert numpy.array_equal(f2.domain, [-1, 1])
+    f2.degree = float(3)
+    assert f2.degree == 3
+    assert isinstance(f2.degree, int)
+
+
+@pytest.mark.parametrize(
+    "cheb_function,degree,answer_key",
+    [
+        (smolyay.basis.ChebyshevFirstKind, 0, {0.5: 1, 1: 1, -1: 1, -0.25: 1, -0.5: 1}),
+        (
+            smolyay.basis.ChebyshevFirstKind,
+            1,
+            {0.5: 0.5, 1: 1, -1: -1, -0.25: -0.25, -0.5: -0.5},
+        ),
+        (
+            smolyay.basis.ChebyshevFirstKind,
+            2,
+            {0.5: -0.5, 1: 1, -1: 1, -0.25: -0.875, -0.5: -0.5},
+        ),
+        (
+            smolyay.basis.ChebyshevSecondKind,
+            0,
+            {0.5: 1, 1: 1, -1: 1, -0.25: 1, -0.5: 1},
+        ),
+        (
+            smolyay.basis.ChebyshevSecondKind,
+            1,
+            {0.5: 1, 1: 2, -1: -2, -0.25: -0.5, -0.5: -1},
+        ),
+        (
+            smolyay.basis.ChebyshevSecondKind,
+            2,
+            {0.5: 0, 1: 3, -1: 3, -0.25: -0.75, -0.5: 0},
+        ),
+    ],
+)
+def test_cheb_call(cheb_function, degree, answer_key):
+    """Test chebyshev polynomial call"""
+    f = cheb_function(degree)
+    assert f(0.5) == pytest.approx(answer_key[0.5])
+    assert f(1) == pytest.approx(answer_key[1])
+    assert f(-1) == pytest.approx(answer_key[-1])
+    assert f(-0.25) == pytest.approx(answer_key[-0.25])
+    assert f(-0.5) == pytest.approx(answer_key[-0.5])
```

Comment:
I'm going to suggest something that goes against my usual advice here. If we turn this into a loop like:
```suggestion
    for x, y in answer_key.items():
        assert f(x) == pytest.approx(y)
```
and you also make the change I suggested above, can we also just add the `Trigonometric` functions to the parameter list for all of these call tests? That would cut down on a lot of code that's copied below.